Migrate PI logic

parisiale opened the pull request.

parisiale committed:
Add StockLocation date checked fields

parisiale committed:
Migrate preceding / next inventory fields

parisiale committed:
Trivial lint fixes

parisiale committed:
Add high level StockLocation PI process logic

The actual scaffolding logic will come later.
Adding trivial unit tests for PI request validation.

ws-mps commented on models/stock_location.py:
```diff
@@ -54,12 +71,113 @@ def get_location(self, location_identifier):
+        self._process_pi_datetime(pi_outcome)
+
+        # @todo: reponse format?
+        return bool(pi_outcome)
```
True?

parisiale committed:
StockLocation PI process expects a singleton

ws-mps commented on models/stock_location.py:
```diff
@@ -54,12 +71,113 @@ def get_location(self, location_identifier):
+        current_time = datetime.now()
+        self.write({'u_date_last_checked': current_time})
+
+        if all([pi_outcome[key] is None
```
If pi_outcome does not contain some of the keys is going to fail.

parisiale replied:
This is equivalent to prior art.

parisiale committed:
PI endpoint returns True on success

parisiale committed:
Fix datetime check for PI update

parisiale committed:
Extend StockInventory::action_done

Adding check on blocked and undone inventories.

parisiale committed:
Extend StockProductionLot by adding get_lot method

parisiale committed:
Rename StockLocation method that executes PI requests

parisiale committed:
First cut of the inventory adjustment logic

parisiale committed:
Improve naming of PI-related things

parisiale committed:
Correct docstring

parisiale committed:
Add docstring

parisiale committed:
Encapsulate PI outcome check

Crating PIOutcome class for improving readability.

parisiale committed:
Add TODO message for improving naming consistency

parisiale committed:
Improving location validation for PI requests

parisiale committed:
Add logic for PI move counts

parisiale committed:
Move StockLocation method for readability

ws-mps commented on models/stock_inventory.py:
```diff
@@ -0,0 +1,33 @@
+
+    @api.multi
+    def action_done(self):
+        self.mapped('location_id').check_blocked(
```
Remove chek_blocked calls, they are already at the blocked_locations module.

ws-mps commented on models/stock_location.py:
```diff
@@ -54,12 +97,274 @@ def get_location(self, location_identifier):
+                    _('You must specify inventory adjustments if you require '
+                      'preceding adjustments.'))
+
+            self._check_obj_locations(keys[:1], request[INVENTORY_ADJUSTMENTS])
```
We should also check the location of the preceding inventory adjustments. Also when there are no preceding inventory adjustments but there is an inventory adjustment the location is not checked.

parisiale replied:
Hmm, not sure I'm following here and I might be missing something.

We're checking the preceding inv adjs; `test03_validate_pi_request_failure_no_adj` covers the case you're talking about.
And there's no location in the `inventory_adjustments` items.
Mind that the outer `location_id` (the mandatory entry of `pi_request`) is already checked, otherwise we won't be here.

parisiale replied:
Ops, my bad, wrong token.

ws-mps commented on models/stock_location.py:
```diff
@@ -54,12 +97,274 @@ def get_location(self, location_identifier):
+            created_pickings += \
+                self._create_pi_count_move_picking(count_move, picking_type_id)
+
+        return created_pickings if created_pickings.exists() else None
```
we don't need the exists

ws-mps commented on models/stock_location.py:
```diff
@@ -54,12 +97,274 @@ def get_location(self, location_identifier):
+            quant_ids = [int(x) for x in count_move['quant_ids']]
+            quants = Quant.browse(quant_ids)
+
+            if len(quants) != len(quant_ids):
```
len(quants) will always be equal to len(quant_ids), we need an exists() either here or at 242.

parisiale replied:
Oh right, good catch. The recordset interface gives you different alternatives.
Perhaps we should write down best practices.

ws-mps commented on models/stock_production_lot.py:
```diff
@@ -0,0 +1,60 @@
+        if isinstance(lot_identifier, int):
+            domain.append(('id', '=', lot_identifier))
+        elif isinstance(lot_identifier, str):
+            domain.append(('name', '=', lot_identifier))
```
This also needs to search by product_id

parisiale replied:
Indeed it has the `product_id` clause.

ws-mps replied:
ah yep, missed that line.

ws-mps commented on models/stock_production_lot.py:
```diff
@@ -0,0 +1,60 @@
+        results = self.search(domain)
+
+        if not results and not no_results:
+            if not create or name is None:
```
we might want to change `name is None` to `not name`, because this will avoid to create a lot with name ''

parisiale committed:
No blocked locations check in StockInventory action_done

parisiale committed:
Fixes for PI internal checks

ws-mps commented on models/stock_location.py:
```diff
@@ -240,11 +241,12 @@ def _create_pi_count_move_picking(self, count_move, picking_type_id):
         elif 'quant_ids' in count_move:
             quant_ids = [int(x) for x in count_move['quant_ids']]
             quants = Quant.browse(quant_ids)
+            num_found_quants = len(list(filter(lambda q: q.exists(), quants)))
```
Can we use len(quants.exists()) please? this will probably be more efficient since it is probably done in one SQL call instead of one per quant

parisiale committed:
Use exists() when looking for found records

parisiale committed:
Add unit tests for PI logic

parisiale committed:
[maint] stock-quant-package return location info

Here we add an extra 'location_id' object to the response
of the endpoint in the title; such object will contain
the package lcoation's id, name, and barcode.

This change is meant to ease the PI front logic.

parisiale committed:
Improve PI tests docstrings

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+                                               create=True)
+
+    def test01_validate_pi_request_success(self):
+        """ No error is raised for a good PI request """
```
`when validating a good PI request` instead of `for ...` ?

Do we have something similar for a request with inventory adjustments?

parisiale replied:
hmmm, `when validating` is superfluous, as you have the method name in the test name...
I prefer to reduce code / comment bloat, but will improve the docstring.

parisiale replied:
Will add an analogous for a an inventory adjustment.

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+            self.test_location_01._validate_perpetual_inventory_request(req)
+
+    def test03_validate_pi_request_failure_no_adj(self):
+        """ Errors if the request contains only preceeding adjustments """
```
I will say contains preceding adjustments but not inventory adjustments.

ws-mps replied:
Also typo: preceeding --> preceding

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+            'location_id': self.test_location_01.id,
+            'pi_count_moves': [
+                {
+                    'quants_ids': [1, 2, 3],
```
How `quants_ids` ended up? quant_ids or quants_ids?

parisiale replied:
it's a typo, it should be `quant_ids`, but it's not used by the method under test

ws-mps replied:
Ok, but can we fix it? This might confuse the reader.

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+        outcome = MockPIOutcome(False)
+        old_dt_checked = self.test_location_01.u_date_last_checked
+        old_dt_success = self.test_location_01.u_date_last_checked_correct
+        time.sleep(1)
```
do we need this sleep?

parisiale replied:
it ensures it will never fail for state

ws-mps replied:
which state? I think by default the dates are empty

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+                        "location_id": self.test_location_01,
+                        "location_dest_id": self.test_location_02.id}]
+
+        pickings = self.test_location_01._process_pi_count_moves(count_moves)
```
this probably needs to be test_location_02._process_pi_.... since we are moving the package from 01 to 02 because we are pi counting 02 and the package is in the wrong location.

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+                         self.package_two.id,
+                         "Picking 1 is not related to the specified package")
+
+    def test09_process_pi_count_moves_valid_request_with_quant(self):
```
Can we have the same test but with 2 quants, different products but in the same package please?

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+        """
+        self.create_quant(self.apple.id, self.test_location_01.id, 4)
+        adjs_req = [{"product_id": self.apple.id,
+                     "package_name": "tes12_NO_PACKAGE",
```
`tes12` is this a typo or intended? same happens later in tes14.

ws-mps replied:
In fact, NO_PACKAGE alone is what we should receive.

parisiale replied:
where's the typo?

parisiale replied:
ah, `tes`, it's just a typo, but it doesn't matter ( you made me recheck the token in the old code...)

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+                        "New package does not seem new")
+        self.assertFalse(stock_info.lot_id, "Wrongly added a lot id")
+
+    def test14_get_stock_drift_new_package_token_multiple_products(self):
```
I would like the same test but different new packages. Also one test with one package multiple products but we only ajdust one of them.

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+        Lot = self.env['stock.production.lot']
+
+        lot_name = "test_15_lot"
+        self.apple.tracking = 'serial'
```
We have products already tracked by serial, no need to change apple.

parisiale replied:
Will change; no strong opinion, but this way is more explicit

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,508 @@
+
+        lot_name = "test_15_lot"
+        self.apple.tracking = 'serial'
+        self.create_quant(self.apple.id, self.test_location_01.id, 7)
```
A quant for a serial tracked item should have the lot_id also set and qty = 1.

parisiale replied:
will create another test, with an existent lot for that

parisiale committed:
PI tests fixes and new tests

parisiale committed:
Trivial naming fixes for PI tests

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -263,7 +298,36 @@ def test09_process_pi_count_moves_valid_request_with_quant(self):
+        self.assertEqual(len(pickings[0].move_line_ids), 2,
+                         "Did not create a picking with 2 move lines")
+
+        for idx in range(2):
```
We can use `for move_line in pickings[0].move_line_ids:` since we don't need idx for anything.

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -394,7 +458,69 @@ def test14_get_stock_drift_new_package_token_multiple_products(self):
+                          package_id=self.package_one.id)
+        adjs_req = [{"product_id": self.apple.id,
+                     "package_name": self.package_one.name,
+                     "quantity": 65}]
```
We already have 65 apples, is this intended?

parisiale replied:
Oh, that should be different than what's in the quant for readability, but it doesn't affect the test.

ws-mps replied:
I agree, but it might add confusion

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -427,10 +552,41 @@ def test15_get_stock_drift_no_package_token_with_lot(self):
+                          lot_id=lot.id)
+        adjs_req = [{"product_id": self.strawberry.id,
+                     "package_name": "test20_NO_PACKAGE_with_existent_lot",
+                     "quantity": 98,
```
Why 98? if it is a serial tracked product this should be 1 or 0. Since we start with 1, we migth want to adjust it to 0.

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,665 @@
+        lot_name = "test_19_lot"
+        adjs_req = [{"product_id": self.strawberry.id,
+                     "package_name": "tes19_NO_PACKAGE_with_new_lot",
+                     "quantity": 78,
```
This quantity should be 1.

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,665 @@
+                     "lot_name": "beautiful_lot"}]
+        inv_adj = self.test_location_01\
+                      ._process_inventory_adjustments(adjs_req)
+        inv = Inventory.search([('name', '=', inv_adj.name)])
```
why search? inv_adj should be alreayd the inventroy adjustment recordset.

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,665 @@
+        package_name = "test22_NO_PACKAGE_with_lot"
+        adjs_req = [{"product_id": self.banana.id,
+                     "package_name": package_name,
+                     "quantity": 28,
```
This should be 1.

parisiale replied:
Why?

ws-mps commented on tests/test_location_pi.py:
```diff
@@ -0,0 +1,665 @@
+        adjs_req = [{"product_id": self.banana.id,
+                     "package_name": package_name,
+                     "quantity": 28,
+                     "lot_name": "beautiful_lot"}]
```
I think banana is not tracked by serial.

parisiale replied:
Indeed, this triggers the `product.tracking` check.

parisiale committed:
Add product tracking vs lot validation for PI requests

parisiale committed:
Extra tests for PI request validation

parisiale committed:
PI tests fixes

parisiale committed:
Fix PI request validation: consider lot tracking

parisiale committed:
Trivial style fix
